### ISSUE TITLE
Recommended usage of web-tree-sitter

### ISSUE THREAD

**Issue by char0n** [tree-sitter/tree-sitter#1580]
Hi everybody,

First off all, thank you for the great library!

With this issue I'd like to clarify the recommended usage of [web-tree-sitter](https://www.npmjs.com/package/web-tree-sitter) which is basically WebAssembly bindings to the Tree-sitter parsing library. The [binding_web/README.md](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/README.md) contains a bunch of examples of how to use the bindings. I have a case where I want to create a library using `web-tree-sitter`. The library will expose a single function called `analyze`.

**browser-patch.ts**:
```ts
import { tail } from 'ramda';
import { isString, isFunction } from 'ramda-adjunct';
// @ts-ignore
import treeSitterWasm from 'web-tree-sitter/tree-sitter.wasm';

// patch fetch() to let emscripten load the WASM file
const realFetch = globalThis.fetch;

if (isFunction(realFetch)) {
  globalThis.fetch = (...args) => {
    // @ts-ignore
    if (isString(args[0]) && args[0].endsWith('tree-sitter.wasm')) {
      // @ts-ignore
      return realFetch.apply(globalThis, [treeSitterWasm, tail(args)]);
    }
    return realFetch.apply(globalThis, args);
  };
}
```

**main.ts**:
```ts
import './browser-patch';

import Parser, { Tree, Language } from 'web-tree-sitter';
import { isPromise } from 'ramda-adjunct';

// @ts-ignore
import treeSitterJson from '../../wasm/tree-sitter-json.wasm';

let jsonLanguage: null | Language | Promise<Language> = null;

/**
 * Lexical Analysis of source string using WebTreeSitter.
 * This is WebAssembly version of TreeSitters Lexical Analysis.
 */
const analyze = async (source: string): Promise<Tree> => {
  // always check that parser has been initialized before used
  await Parser.init();
  const parser = new Parser();

  // load language if not loaded
  if (jsonLanguage === null) {
    jsonLanguage = Parser.Language.load(treeSitterJson);
  }

  // resolve language if not yet resolved
  if (isPromise(jsonLanguage)) {
    try {
      // resolve the loaded language
      jsonLanguage = await jsonLanguage;
    } catch (e) {
      // allow next analyze call to load the language again if it failed before
      jsonLanguage = null;
      throw e;
    }
  }

  parser.setLanguage(jsonLanguage);

  return parser.parse(source);
};

export default analyze;
```

I want `analyze` function to be as lazy as it can be. I understand that it means that first parse is going to be slow, as both the `tree-sitter.wasm` and `tree-sitter-json.wasm` needs to be loaded before actual parsing. I'm not sure if this is the way how `web-tree-sitter` was intended to be used. First of all I always call `await Parser.init();` inside `analyze` function before anything else, as I need to be sure that the library is initialized before used. Next JSON grammar is loaded when the `analyze` is called for the first time. New instance of `Parser` class is created every time the `analyze` function is called and is provided loaded JSON grammar. 

The laziness can be switched to eagerness by calling `await Parser.init();` and `Parser.Language.load(treeSitterJson)` outside of `analyze` function body (that's clear). But what about creating new parser instance every time we need to parse something? The usual Modus operandi with other parsers that I've seen in past is that for every parsing new parser instance is created to avoid possible parser instance state pollution. Is this the same case with `web-tree-sitter` (state in parser instance) or is it absolutely safe to reuse single parser instance because it has no internal state?

Thank you for any clarification. I can issue a PR for the [binding_web/README.md](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/README.md) file if we manage to clarify my questions here.

**Comment by amaanq** [comment]
Yeah, to answer your question, you only need to call `Parser.init()` and `Parser.Language.load(...)` once. The `init` function initializes the wasm module with with a few setup functions called to configure some internal state stuff, but you don't have to do this multiple times, and the `load` function loads the wasm file and constructs a new language from it, you don't have to load this everytime you want to use the language.

**Comment by char0n** [comment]
Hi @amaanq,

Thank you! There is still an open question of necessity of `const parser = new Parser();` - do I need to create parser instance before every parsing or am I just fine with one instance parsing many source strings?

**Comment by amaanq** [comment]
oh yeah, sorry i missed that - you can definitely reuse a parser, but if you want to change the language used, you'd have to call `setLanguage`
